### DISCUSSION TITLE
parallax scrolling / multi-layer support

### DISCUSSION THREAD

**Discussion by msemegen** [erincatto/box2d#831] [Q&A] [answered]
I'm trying to implement parallax scrolling with multi-layer support, where each layer uses Box2D for collision detection and physics. I know how to separate collision detection between layers (using masks), but I’m unsure about handling movement and scaling. Objects on layers further from the camera should move more slowly and appear smaller. How can I make Box2D account for this?

**Comment by erincatto** [comment, marked as answer]
I would use a larger visual viewport for the layers further back. Make it a rendering problem. Also you might consider using a separate Box2D world for each layer instead of using masks.

**Reply by msemegen** [reply]
Having a separate Box2D world for each layer is an interesting suggestion.

**Comment by notnullnotvoid** [comment]
For the game I'm making, we wanted something similar with physics happening on multiple layers/planes. I figured I'd write up something here about how I solved it, since I thought that might be of interest.

For us, having separate worlds for each layer wasn't ideal because we wanted to be able to have objects that span multiple layers (so e.g. an object spanning layers 1 and 2 would collide with both layer-1 objects and layer-2 objects, but layer-1 and layer-2 objects wouldn't collide with each other) as well as being able to joint objects on different layers together.

Using collision masks wouldn't work either because we wanted to have a large number of layers, each of which can have objects on it with all the same physics interactions as on any other layer. As far as I can tell this would require essentially duplicating all the collision mask bits we use per each layer, but `b2Filter::categoryBits` is only a 64-bit int, so if `numUsedCollisionCategories * numUsedPhysicsLayers > 64` then you're out of luck with this method, which would definitely be the case for us.

I ended up making a small modification to Box2D - basically giving `b2Filter` a specified range of layers that it occupies, and ignoring collision between shapes that don't have any layers in common. It's only a few lines of code changed but I found it tremendously useful: https://github.com/notnullnotvoid/box2d-layers/commit/9ca8391c46a3a6d90126d3fe115327ba04709e38 You do have to account for this in user code when doing spatial queries (overlap tests / ray casts / shape casts) where applicable, but that's already the case for things like `b2Filter::groupIndex` and `collideConnected` on joints, so it's no more troublesome than those. Also, it's possible I missed some places in the code that should be changed to account for the new properties, but this was at least enough for how we're using it.

Alternatively, you could use a bitmask for layers instead of a range. Then you'd be able to specify a non-contiguous set of layers for each shape, at the cost of having fewer total layers for the same memory footprint. A single `uint64_t` would probably fine for most things - our use case of wanting more than 64 total layers is kind of extreme.

As for rendering perspective/parallax, we handle that entirely separately from Box2D. One thing that's a bit lacking about our approach is that the debug rendering via `b2DebugDraw` doesn't obey perspective in any particular way, which can get confusing. I might look into improving that at some point but for now I'm just living with it.
